### ISSUE TITLE
Multi-monitor Idea - Re-apply rules

### ISSUE THREAD

**Issue by danielcherubini** [baskerville/bspwm#1135]
Hi.

I've been trying to find a good way to reapply rules for nodes. This is handy for two things.

1. Hotswaping monitors
2. Having bound to a key

So lets look at the first usecase, I have my `bspwmrc` setup so that it detects a single monitor, or two monitors and it uses an order to determine the primary and secondary monitor in the case it has dual monitors.

https://gitlab.com/danielcherubini/dotfiles/-/blob/master/bspwm/bspwmrc

```bash
declare -a monitorarr
monitors=0
for m in $(xrandr --listactivemonitors | grep "DP" | cut -d " " -f6); do
	monitorarr[$monitors]=$m
	((monitors++))
done

case "$monitors" in
	"1" )
		echo "1"
		echo ${monitorarr[0]}
		bspc monitor -d I II III IV V VI VII VIII IX X
		;;
	"2" )
		echo "2"
		bspc monitor ${monitorarr[0]} -d I II III IV V
		bspc monitor ${monitorarr[1]} -d VI VII VIII IX X
		;;
esac
```

This is working fantastic..  however lets say I plug my laptop into my screen, I can either use one of the two USB-C ports, so with this config above, it doesn't matter which one I use.

I activate the second monitor, either extending the displays, or using it as the only display...

Now the monitor is empty, all the displays are attached to another monitor. So, obviously I want to grab each display, and move them to the new monitor, so for that I wrote a little script.. 

https://gitlab.com/danielcherubini/dotfiles/-/blob/master/bspwm/monitor.sh

```bash
#!/usr/bin/env bash

# move $2 to $1

for desktop in $(bspc query -D -m $1)
do
  bspc desktop $desktop --to-monitor $2
done


bspc query -D -m --names
bspc monitor $1 -r
bspc desktop Desktop -r
```

It's not perfect, but it gets the job done.

I trigger it with another script which I've hotkeyed to the "change screen" button on my keyboard

https://gitlab.com/danielcherubini/dotfiles/-/blob/master/bspwm/swap-monitors.sh

This works .. sometimes.. there's issues with it. Sometimes it loses nodes.

It made me think, this is the wrong way off doing this... the right way would be something inside BSPWM where I could trigger via `bspc` to "restart the wm and re-apply the rules" 

basically something like this

1. get all the monitors and then remove them
2. restart bspwm so it runs the bspwmrc file creates new monitors
3. use the re-apply the rules command so all the nodes that match the rules will move to the right desktops.

Thoughts?

**Comment by emanuele6** [comment]
> I've been trying to find a good way to reapply rules for nodes.

You can use `xdo hide <wid> && xdo show <wid>` to "reapply rules".

Example: script that re-applies rules to all windows. 
```bash
#!/bin/env bash

for wid in $(bspc query -N -n .window); do
    xdo hide "$wid" && xdo show "$wid"
done
```

--------------------------------------

[xdo](https://github.com/baskerville/xdo.git)
